sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala

```diff
@@ -592,3 +593,28 @@ case class QualifyLocationWithWarehouse(catalog: SessionCatalog) extends Rule[Lo
       c.copy(tableDesc = newTable)
   }
 }
+
+object CollationCheck extends (LogicalPlan => Unit) {
+  def apply(plan: LogicalPlan): Unit = {
+    plan.foreach {
+      case operator: LogicalPlan =>
+        operator transformExpressionsUp {
+          case e@(_: Collation | _: Collate) =>
+            if (!SQLConf.get.collationEnabled) {
+              throw QueryCompilationErrors.collationNotEnabledError()
+            }
+            e
+          case other =>
+            if (other.children.exists(hasCollationExpression)) {
+              if (!SQLConf.get.collationEnabled) {
+                throw QueryCompilationErrors.collationNotEnabledError()
+              }
+            }
+            other
+        }
+    }
+  }
+
+  private def hasCollationExpression(expression: Expression): Boolean =
```

Comment:
I think it should be `isCollationExpression`, as it checks the given expression itself, not its children.